### ISSUE TITLE
Realm 0.15.3 w/ Jest unable to find module realm.node

### ISSUE THREAD

**Issue by csotiriou** [realm/realm-js#789]
Hello,
After upgrading  to realm 0.15.3, my Jest unit tests stopped working. I have tried both npm and Yarn to install my node modules, but still, it seems that realm.node cannot be found. My app in React Native runs correctly.

It seems similar to this: #713 but I have also tried npm to no avail.

```
./node_modules/realm/compiled/node-v48_darwin_x64/realm.node' from 'index.js'
      
      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:150:17)
      at Object.<anonymous> (node_modules/realm/lib/index.js:77:28)
      at Object.<anonymous> (backend/db/util/db.util.js:5:12)

```

EDIT: It was working just fine with both yarn and npm in Realm 0.15.0
EDIT2:

After some time debugging, I think I have a good idea why this happens.

I believe the problem lies in the script "install.js"

It seems that in order to install realm.node, the script checks if it is in a react-native environment, by trying to require the react-native package.

The thing is, that when running unit-tests, we are not inside a react-native environment, we are in the node environment even if the react native package is present, since the unit tests are ran in the host machine. Therefore, ```node-pre-gyp install --fallback-to-build``` will never run if we are developing in react native.

Bottom line: When trying to have unit tests in react-native, they will always fail, because realm.node will never be installed.

EDIT3: Temporary fix:
I added a post install script in the package.json:
```    "postinstall" : "./bin/build-realm.sh" ```

and inside the folder ./bin, I created a `build-realm.sh` file with the following contents:

```
#!/usr/bin/env bash

set -e

npm install -g node-gyp
npm install -g node-pre-gyp

cd ./node_modules/realm/

node-pre-gyp install --fallback-to-build
```

This will install both node-gyp and node-pre-gyp before executing the build command directly. In my case, I will add the check for pre-existing node-gyp and node-pre-gyp binaries, but you get the idea. It works now.

However, in the case of realm 0.15.3, I believe that no matter what unit testing library one uses, no unit tests will be able to run. I believe that the realm.node file should be built / downloaded immediately after installing, regardless of whether we are developing for react native or not, since unit tests are ran by node, not react native.

**Comment by csotiriou** [comment]
After some time debugging, I think I have a good idea why this happens.

I believe the problem lies in the script "install.js"

It seems that in order to install realm.node, the script checks if it is in a react-native environment, by trying to require the react-native package. 

The thing is, that when running unit-tests, we are not inside a react-native environment, we are in the node environment even if the react native package is present, since the unit tests are ran in the host machine. Therefore, `node-pre-gyp install --fallback-to-build` will never run if we are developing in react native.

Bottom line: When trying to have unit tests in react-native, they will always fail, because realm.node will never be installed.

**Comment by kristiandupont** [comment]
@csotiriou, you are right, that is exactly what happens: as we detect you are developing for React Native, we don't install the Node binaries. The reason is partly to cut down installation time but mostly because we don't yet support Node on Windows, so installing Realm on a Windows machine (even if developing for React Native) would fail.

But this is obviously not the way to go as we still need the node binaries for unit testing, so I will have to modify the install script further.  Thank you for reporting this!

**Comment by csotiriou** [comment]
Thanks for the quick resolution!

Until realm-js 0.15.4 is released, I will use the workaround.

**Comment by kristiandupont** [comment]
@csotiriou the new version is out. Try and see if that helps.

**Comment by csotiriou** [comment]
@kristiandupont It works. thank you.

**Comment by j2kun** [comment]
I am working on a project using realm 0.15.4 and I experienced this bug, and the workaround did indeed fix it. Are you certain this was fixed? The problem could be that I had not installed node-pre-gyp before `npm install`ing the project, but I didn't see any error messages...

**Comment by fealebenpae** [comment]
@j2kun realm has an internal dependency on `node-pre-gyp` and that should be installed automatically. Do you use `npm@2` and not a newer version, by any chance?

**Comment by j2kun** [comment]
I saw that node-gyp was installed in my project's `node_modules`, but I still encountered the error above. What fixed it for me was to install it with `-g`.

(I apologize, I am quite new to node, and I am still learning how all these CLI tools and packages are related)

**Comment by fealebenpae** [comment]
@j2kun I think our `scripts/install.js` which invokes `node-pre-gyp` expects the `npm@3` behavior where `realm` and `node-pre-gyp` will be adjacent to each-other in your `node_modules` folder. `npm@2` creates deep hierarchies, so that `node-pre-gyp`, as a dependency of realm, will be placed in `node_modules/realm/node_modules` instead of in `node_modules`.

This doesn't seem to be the problem in your case, though - we'll keep investigating.

**Comment by Cody-G-G** [comment]
I'm using realm@1.0.2 with react-native, npm@4.1.1, node@v6.4.0, on Windows. I am encountering the same issue detailed in the OP. I see mentioned that Node on Windows is not supported yet, but I was wondering whether there would be any workaround for it?

**Comment by fealebenpae** [comment]
Yes, this is precisely the issue - Jest runs tests as Node.js code, and Realm is not supported in Node.js on Windows just yet, but it's something we're working on at the moment.
